lib/WUI/nhttp/send_json.cpp

```diff
@@ -0,0 +1,128 @@
+#include "send_json.h"
+#include "handler.h"
+#include "headers.h"
+
+#include <timing.h>
+#include <segmented_json_macros.h>
+#include <http/chunked.h>
+#include <transfers/monitor.hpp>
+
+using transfers::Monitor;
+using namespace json;
+using namespace http;
+
+namespace nhttp::handler {
+
+json::JsonResult EmptyRenderer::renderState(size_t resume_point, json::JsonOutput &output, Empty &state) const {
+    return generator(resume_point, output);
+}
+
+JsonResult TransferRenderer::renderState(size_t resume_point, json::JsonOutput &output, TransferState &state) const {
+    // Note: We are guaranteed to get a status, because we checked if it is present
+    // before calling this and we set allow_stale to true
+    auto transfer_status = Monitor::instance.status(true);
+    // Keep the indentation of the JSON in here!
+    // clang-format off
+    JSON_START;
+        JSON_OBJ_START
+            if (transfer_status->id != state.transfer_id) {
+                // if transfer changes mid report, bail out
+                transfer_status.reset();
+            }
+
+            JSON_FIELD_STR("target", "usb") JSON_COMMA;
+            JSON_FIELD_STR_FORMAT_G(transfer_status.has_value(), "type", "%s", to_str(transfer_status->type)) JSON_COMMA;
+            JSON_FIELD_STR_FORMAT_G(transfer_status.has_value(), "size", "%d", transfer_status->expected) JSON_COMMA;
+            //FIXME: Right now should be seconds from epoch start, would be much nicer
+            // for us to make it seconds from print start and make the client compute the rest,
+            // also it would solve some potential problems with NTP and time zones etc.
+            JSON_FIELD_STR_FORMAT_G(transfer_status.has_value(), "start_time", "%d", time(nullptr) - (ticks_s() - transfer_status->start)) JSON_COMMA;
+            JSON_FIELD_BOOL_G(transfer_status.has_value(), "to_print", transfer_status->print_after_upload) JSON_COMMA;
+            // Note: This works, because destination cannot go from non null to null
+            // (if one transfer ends and another starts mid report, we bail out)
+            if (transfer_status->destination) {
```

Comment:
Not really trivial to see that we don't need `if (transfer_status.has_value() && transfer_status->destination != nullptr)`… (though it is true).